pkg/yurthub/healthchecker/health_checker.go

```diff
@@ -119,156 +93,129 @@ func (hcm *healthCheckerManager) healthzCheckLoop(stopCh <-chan struct{}) {
-	return false
 }
 
-func (hcm *healthCheckerManager) UpdateLastKubeletLeaseReqTime(lastKubeletLeaseReqTime time.Time) {
-	hcm.Lock()
-	defer hcm.Unlock()
-	hcm.lastKubeletLeaseReqTime = lastKubeletLeaseReqTime
-}
+func (hc *cloudAPIServerHealthChecker) IsHealthy() bool {
+	for _, prober := range hc.probers {
+		if prober.IsHealthy() {
+			return true
+		}
+	}
 
-func (hcm *healthCheckerManager) getLastKubeletLeaseReqTime() time.Time {
-	hcm.RLock()
-	defer hcm.RUnlock()
-	return hcm.lastKubeletLeaseReqTime
+	return false
 }
 
-func newChecker(
-	cfg *config.YurtHubConfiguration,
-	tp transport.Interface,
-	url *url.URL,
-	setLastNodeLease setNodeLease,
-	getLastNodeLease getNodeLease,
-) (*checker, error) {
-	restConf := &rest.Config{
-		Host:      url.String(),
-		Transport: tp.CurrentTransport(),
-		Timeout:   time.Duration(cfg.HeartbeatTimeoutSeconds) * time.Second,
-	}
-	kubeClient, err := clientset.NewForConfig(restConf)
-	if err != nil {
-		return nil, err
-	}
-
-	nl := NewNodeLease(kubeClient, cfg.NodeName, defaultLeaseDurationSeconds, cfg.HeartbeatFailedRetry)
-	c := &checker{
-		nodeLease:        nl,
-		lastTime:         time.Now(),
-		clusterHealthy:   false,
-		healthyThreshold: cfg.HeartbeatHealthyThreshold,
-		healthyCnt:       0,
-		remoteServer:     url,
-		onFailureFunc:    tp.Close,
-		setLastNodeLease: setLastNodeLease,
-		getLastNodeLease: getLastNodeLease,
+// BackendHealthyStatus returns the healthy stats of specified server
+func (hc *cloudAPIServerHealthChecker) BackendHealthyStatus(server *url.URL) bool {
+	if prober, ok := hc.probers[server.String()]; ok {
+		return prober.IsHealthy()
 	}
-	return c, nil
+	// If there is no checker for server, default unhealthy.
+	return false
 }
 
-func (c *checker) check() bool {
-	baseLease := c.getLastNodeLease()
-	lease, err := c.nodeLease.Update(baseLease)
-	if err == nil {
-		if err := c.setLastNodeLease(lease); err != nil {
-			klog.Errorf("set last node lease fail: %v", err)
-		}
-		c.healthyCnt++
-		if !c.isHealthy() && c.healthyCnt >= c.healthyThreshold {
-			c.setHealthy(true)
-			now := time.Now()
-			klog.Infof("cluster becomes healthy from %v, unhealthy status lasts %v, remote server: %v", now, now.Sub(c.lastTime), c.remoteServer.String())
-			c.lastTime = now
-			metrics.Metrics.ObserveServerHealthy(c.remoteServer.Host, 1)
+func (hc *cloudAPIServerHealthChecker) run(stopCh <-chan struct{}) {
+	intervalTicker := time.NewTicker(time.Duration(hc.heartbeatInterval) * time.Second)
+	defer intervalTicker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			klog.Infof("exit normally in health check loop.")
+			return
+		case <-intervalTicker.C:
+			// Ensure that the node heartbeat can be reported when there is a healthy remote server.
+			// Try to detect all remote server in a loop, if there is a remote server can update nodeLease, exit the loop.
+			for i := 0; i < len(hc.remoteServers); i++ {
+				p := hc.getProber()
+				if p.Probe(ProbePhaseNormal) {
+					break
+				}
```

Comment:
Why do we use such a complex way to probe all remote servers? 
Could we just 
```go
for _, p := range hc.probers {
  if p.Probe(ProbePhaseNormal) {
    break
  }
}
```

Reply:
in the next cycle, we need to probe from the `last+1` remote server instead of the first remote server.

Reply:
because we need to make sure that only one lease is updated to kube-apiserver, at the same time, we can probe the healthy status for all remote servers.

Reply:
sounds good